### PR TITLE
Energy-dependent prompt and recoverable fission Q-value tallies.

### PR DESCRIPTION
**Pull request by smharper**
This PR allows us to use the ENDF MT=1 MF=458 data in a tally.  This tells us how much of the fission energy is carried by the fragments, the prompt neutrons, the prompt gammas, the delayed neutrons etc.  It is also a function of incident neutron energy.

### Example results

For now, I've just added two tallies, `q-fission-prompt` and `q-fission-recoverable`, but if desired we can make a tally for each species of outgoing fission particles.  It will probably be handy for the purposes of depletion to be able to separate the prompt energy production from the delayed stuff that comes from later neutron and beta decays.  It's nice to have the energy dependence, but it won't matter much for reactor calculations.  The Q-value decreases about one-to-one with increasing neutron energy so the difference between fast and thermal reactors is only 0--1 MeV per fission.

I ran some example problems to see what the Q-values for the big three actinides look like at different energies.  I used a BEAVRS pincell with some Pu sprinkled in for a thermal spectrum.  For a fast spectrum, I got an SFR pincell from the nice folks here at ANL NE.  I also got a high energy spectrum by voiding a BEAVRS pincell and running it with a 14.1 MeV fixed source with vacuum BC's.  Here's what the Q-values look like at these different energies (the listed tally score is divided by a fission tally to get the Q-value rather than a power)
```
nuclide                 score              LWR              SFR              D-T
   U235         kappa-fission  193.41 +/- 0.14  193.41 +/- 0.08  193.41 +/- 0.08
   U235      fission-q-prompt  180.55 +/- 0.13  180.30 +/- 0.07  168.75 +/- 0.07
   U235  fission-q-recoverabl  193.39 +/- 0.14  193.09 +/- 0.08  179.60 +/- 0.08

   U238         kappa-fission  197.79 +/- 0.11  197.79 +/- 0.12  197.79 +/- 0.09
   U238      fission-q-prompt  178.29 +/- 0.10  178.51 +/- 0.11  169.55 +/- 0.07
   U238  fission-q-recoverabl  194.57 +/- 0.11  194.82 +/- 0.12  184.23 +/- 0.08

  Pu239         kappa-fission  198.90 +/- 0.15  198.90 +/- 0.07  198.90 +/- 0.08
  Pu239      fission-q-prompt  188.41 +/- 0.14  188.01 +/- 0.07  175.01 +/- 0.07
  Pu239  fission-q-recoverabl  198.89 +/- 0.15  198.43 +/- 0.07  183.52 +/- 0.08
```

As you can see, the difference between fast and thermal is only about 200--500 keV.  Strangely, there is a huge difference between `fission-q-recoverable` and `kappa-fission` for U238.  I'll try to figure out why we have that difference before the PR is merged.  (As a side note, the above uncertainties show a good example of when covariances matter.  Check out the values and uncertainties on the kappa-fissions across the energies.)

### The data

Unfortunately, this data isn't available in ACE files.  We just got a shiny new HDF5 data format so we can stick the 458 data in there.  But this means that the PythonAPI needs to be able to read the data from ENDF.  I added that capability and tested it with ENDF/B-VII.1 and JEFF-3.2.  But I think it would be mean to make our users download an entire ENDF neutron sublibrary just for the 458 data so I packaged the relevant data from ENDF/B-VII.1 into an HDF5 file and stuck it in the data directory.  The `get_nndc_data.py` script will automatically read that HDF5 file and append the 458 data to our HDF5 cross sections.  Unfortunately all this data handling really ballooned the LoC in this PR beyond what you'd expect.

I used two different HDF5 formats.  I have reasons for doing this, but I'm not 100% convinced it's the best way to go.  The "compact" format that I used to store the ENDF/B-VII.1 data is basically just a direct translation from ENDF to HDF5.  It's just a set of coefficients with all of the components still separate.  The cross sections HDF5, however, uses data that's been summed up into prompt and recoverable components, stripped of uncertainties, and sometimes converted into tables.

The reason I think those two formats should remain separate is because the "compact" one will not have to change until the ENDF format does, whereas we might change the one in our usual HDF5 files when we decide we to add new tallies or refactor the way we represent `Function1D`s.  Of course, the downside is that it's rather confusing having two different formats.

One small note: some 458 data requires nu.  I've only implemented this functionality for nu values that are stored as a `Tabluated1D` although they can be stored as a `Polynomial`.  There are only a few evaluations in ENDF/B-VII.1 that use polynomial nu (just Ra if I remember correctly), and those nuclides don't have 458 data available.  Same must be true of JEFF-3.2 since the Python API didn't error out when I tried it.

@paulromano: we'll need to update the data in your box for Travis to pass this.  You can be my first usability tester :)

### CODE REVIEW

**Comment by smharper** [comment]
I figured out the difference between `kappa-fisison` and `fission-q-recoverable` with U238.  Since the U238 cross section is small at low energies, U238 produces all of its power from high energy incident neutrons where the energy-dependence starts to matter.  This result is super easy to produce with our new data processing tools:
```python
fname = '/home/smharper/openmc/data/nndc_hdf5/U238_71c.h5'
data238 = openmc.data.IncidentNeutron.from_hdf5(fname)

fname = '/home/smharper/openmc/data/nndc_hdf5/U235_71c.h5'
data235 = openmc.data.IncidentNeutron.from_hdf5(fname)

# Use the U235 fission spectrum as the incident neutron flux
fiss_spect = data235[18].products[0].distribution[0].energy.energy_out[0]

E = fiss_spect.x
flux = fiss_spect.p * np.sqrt(E)

rxn_rate = flux * data238[18].xs(E)
power = rxn_rate * data238.fission_energy.q_recoverable(E)

avg_Q = np.trapz(power, E) / np.trapz(rxn_rate, E)
print(avg_Q)
```
That code gives me a Q-value of 194.13 MeV, very close to the OpenMC results above.

File: openmc/data/endf_utils.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+.. _Pyne: http://www.pyne.io
+
+"""
+
+import re
+
+def read_float(float_string):
+    """Parse ENDF 6E11.0 formatted string into a float."""
+    assert len(float_string) == 11
+    pattern = '([\s\\-]\d+\\.\d+)([\\+\\-]\d+)'
```
Comment (paulromano) [inline openmc/data/endf_utils.py]: Use a raw string for regular expressions. Also, ENDF files are not guaranteed to (and in fact often do not) use exponential form.

File: openmc/data/fission_energy.py [hunk 2]
```diff
@@ -0,0 +1,627 @@
...
+        if error_present:
+            for coeffs in value.values(): coeffs[2] *= 1e-6
+            for coeffs in uncertainty.values(): coeffs[2] *= 1e-6
+
+        # Perform the sanity check again... just in case.
+        for coeffs in value.values():
+            second_order = coeffs[2]
+            if abs(second_order) * 1e12 > 1e8:
+                raise ValueError("Encountered a ludicrously large second-"
+                                 "order polynomial coefficient.")
```
Comment (paulromano) [inline openmc/data/fission_energy.py]: Is this ever expected to happen?

File: openmc/data/fission_energy.py [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+        Name of the output HDF5 file.  Default is 'fission_Q_data.h5'.
+    comment : str
+        Comment to write in the output HDF5 file.  Defaults to no comment.
+    verbose : bool
+        If True, print the name of each isomer as it is read.  Defaults to
+        False.
+
+    """
+    # Open the output file.
+    if output_name is None: output_name = 'fission_Q_data.h5'
```
Comment (paulromano) [inline openmc/data/fission_energy.py]: Put the default in the method signature

File: openmc/data/fission_energy.py [hunk 4]
```diff
@@ -0,0 +1,627 @@
...
+        for i in range(len(coeffs)):
+            coeffs[i] *= 10**(-6 + 6*i)
+    for coeffs in uncertainty.values():
+        for i in range(len(coeffs)):
+            coeffs[i] *= 10**(-6 + 6*i)
+
+    return value, uncertainty
+
+
+def write_compact_458_library(endf_files, output_name=None, comment=None,
```
Comment (paulromano) [inline openmc/data/fission_energy.py]: Is there a reason you didn't make this a method of `FissionEnergyRelease`?

File: openmc/data/fission_energy.py [hunk 5]
```diff
@@ -0,0 +1,627 @@
...
+        data.append(read_float(lines[2 + row][11*column:11*(column+1)]))
+
+    # Declare the coefficient names and the order they are given in.  The LIST
+    # contains a value followed immediately by an uncertainty for each of these
+    # components, times the polynomial order + 1.
+    labels = ('EFR', 'ENP', 'END', 'EGP', 'EGD', 'EB', 'ENU', 'ER', 'ET')
+
+    # Associate each set of values and uncertainties with its label.
+    value = dict()
+    uncertainty = dict()
```
Comment (paulromano) [inline openmc/data/fission_energy.py]:
```Python
value = {}
unceratinty = {}
```

File: openmc/data/fission_energy.py [hunk 6]
```diff
@@ -0,0 +1,627 @@
...
+    for i in range(len(labels)):
+        value[labels[i]] = data[2*i::18]
+        uncertainty[labels[i]] = data[2*i + 1::18]
```
Comment (paulromano) [inline openmc/data/fission_energy.py]:
```Python
for i, label in enumerate(labels):
    value[label] = ...
    ...

File: src/tally.F90 [hunk 7]
```diff
@@ -703,14 +703,14 @@ subroutine score_general_ce(p, t, start_index, filter_index, i_nuclide, &
...
-                     micro_xs(p%event_nuclide)%fission / &
-                     micro_xs(p%event_nuclide)%absorption * flux
+            ! fission scaled by kappa-fission
+            associate (nuc => nuclides(p % event_nuclide))
+              if (micro_xs(p % event_nuclide) % absorption > ZERO .and. &
+                   nuc % fissionable) then
+                score = p % absorb_wgt * &
+                     nuc % reactions(nuc % index_fission(1)) % Q_value * &
+                     micro_xs(p % event_nuclide) % fission / &
+                     micro_xs(p % event_nuclide) % absorption
```
Comment (paulromano) [inline src/tally.F90]: flux disappeared here -- is that intended?

File: src/tally.F90 [hunk 8]
```diff
@@ -719,35 +719,36 @@ subroutine score_general_ce(p, t, start_index, filter_index, i_nuclide, &
...
-                score = p%last_wgt * &
-                     nuc%reactions(nuc%index_fission(1))%Q_value * &
-                     micro_xs(p%event_nuclide)%fission / &
-                     micro_xs(p%event_nuclide)%absorption * flux
+            associate (nuc => nuclides(p % event_nuclide))
+              if (nuc % fissionable) then
+                score = p % last_wgt * &
+                     nuc % reactions(nuc % index_fission(1)) % Q_value * &
+                     micro_xs(p % event_nuclide) % fission / &
+                     micro_xs(p % event_nuclide) % absorption
```
Comment (paulromano) [inline src/tally.F90]: ditto about flux

File: openmc/data/fission_energy.py [hunk 9]
```diff
@@ -0,0 +1,627 @@
...
+        # Skip non-fissionable nuclides.
+        if not ident['LFI']: continue
+
+        # Get the important bits.
+        data = _extract_458_data(fname)
+        if data is None: continue
+        value, uncertainty = data
+
+        # Make a group for this isomer.
+        name = str(ident['Z']) + str(ident['A'])
```
Comment (paulromano) [inline openmc/data/fission_energy.py]: Maybe `ATOMIC_SYMBOL(ident['Z'])` instead? Would make it consistent with our other HDF5 data

File: openmc/data/fission_energy.py [hunk 10]
```diff
@@ -0,0 +1,627 @@
...
+        components = [s.decode() for s in fin.attrs['component order']]
+
+        nuclide_name = str(incident_neutron.atomic_number)
+        nuclide_name += str(incident_neutron.mass_number)
+        if incident_neutron.metastable != 0:
+            nuclide_name += '_m' + str(incident_neutron.metastable)
+
+        if nuclide_name not in fin: return None
+
+        data = {c : fin[nuclide_name + '/data'][i, 0, :]
```
Comment (paulromano) [inline openmc/data/fission_energy.py]: No space before colon, `data = {c: ...}`

File: openmc/data/fission_energy.py [hunk 11]
```diff
@@ -0,0 +1,627 @@
...
+                nu_prompt = [p for p in incident_neutron[18].products
+                             if p.particle == 'neutron'
+                             and p.emission_mode == 'prompt']
+            elif 19 in incident_neutron.reactions:
+                nu_prompt = [p for p in incident_neutron[19].products
+                             if p.particle == 'neutron'
+                             and p.emission_mode == 'prompt']
+            else:
+                raise ValueError('IncidentNeutron data has no fission '
+                                  'reaction.')
```
Comment (paulromano) [inline openmc/data/fission_energy.py]: hanging indent off by one space

File: openmc/data/fission_energy.py [hunk 12]
```diff
@@ -0,0 +1,627 @@
...
+        obj.betas = Polynomial(group['betas'].value)
+        obj.neutrinos = Polynomial(group['neutrinos'].value)
+
+        if group.attrs['format'].decode() == 'Madland':
+            obj.form = 'Madland'
+            obj.prompt_neutrons = Polynomial(group['prompt_neutrons'].value)
+        elif group.attrs['format'].decode() == 'Sher-Beck':
+            obj.form = 'Sher-Beck'
+            obj.prompt_neutrons = Tabulated1D.from_hdf5(
+                                                       group['prompt_neutrons'])
```
Comment (paulromano) [inline openmc/data/fission_energy.py]: Hanging indents usually four spaces

File: openmc/data/endf_utils.py (hunk 1 above)
Comment (smharper) [inline openmc/data/endf_utils.py]: Okay, I'm using `re.sub` which will just return the unmodified input float if the data isn't exponential form.

File: openmc/data/fission_energy.py (hunk 2 above)
Comment (smharper) [inline openmc/data/fission_energy.py]: It is not so I've removed the check

File: openmc/data/fission_energy.py (hunk 4 above)
Comment (smharper) [inline openmc/data/fission_energy.py]: This function doesn't use `FissionEnergyRelease` data or methods so I don't think it should be a method.

File: src/tally.F90 (hunk 7 above)
Comment (smharper) [inline src/tally.F90]: Not intended.  Fixed

File: openmc/data/fission_energy.py (hunk 9 above)
Comment (smharper) [inline openmc/data/fission_energy.py]: Good point.  Done.

**Pull request closed without merging**